FAILED: HighLevelOutgoingConnectionSpec

```
[ERROR] [01/18/2016 10:25:47.030] [HighLevelOutgoingConnectionSpec-akka.actor.default-dispatcher-4] [akka.actor.ActorSystemImpl(HighLevelOutgoingConnectionSpec)] Outgoing request stream error
java.lang.IllegalArgumentException: requirement failed: Cannot pull closed port (method)
	at scala.Predef$.require(Predef.scala:219)
	at akka.stream.stage.GraphStageLogic.pull(GraphStage.scala:419)
	at akka.http.impl.engine.client.OutgoingConnectionBlueprint$ResponseParsingMerge$$anon$3$$anonfun$1.apply$mcV$sp(OutgoingConnectionBlueprint.scala:207)
	at akka.stream.stage.GraphStageLogic$Emitting.followUp(GraphStage.scala:830)
	at akka.stream.stage.GraphStageLogic$EmittingSingle.onPull(GraphStage.scala:879)
	at akka.stream.impl.fusing.GraphInterpreter.processEvent(GraphInterpreter.scala:593)
	at akka.stream.impl.fusing.GraphInterpreter.execute(GraphInterpreter.scala:533)
	at akka.stream.impl.fusing.GraphInterpreterShell.runBatch(ActorGraphInterpreter.scala:445)
	at akka.stream.impl.fusing.GraphInterpreterShell.receive(ActorGraphInterpreter.scala:400)
	at akka.stream.impl.fusing.ActorGraphInterpreter$$anonfun$receive$1.applyOrElse(ActorGraphInterpreter.scala:549)
	at akka.actor.Actor$class.aroundReceive(Actor.scala:480)
	at akka.stream.impl.fusing.ActorGraphInterpreter.aroundReceive(ActorGraphInterpreter.scala:495)
	at akka.actor.ActorCell.receiveMessage(ActorCell.scala:526)
	at akka.actor.ActorCell.invoke(ActorCell.scala:495)
	at akka.dispatch.Mailbox.processMailbox(Mailbox.scala:257)
	at akka.dispatch.Mailbox.run(Mailbox.scala:224)
	at akka.dispatch.Mailbox.exec(Mailbox.scala:234)
	at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
	at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
	at scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
	at scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)

[info] - should catch response stream truncation *** FAILED *** (29 milliseconds)
[info]   Expected exception akka.stream.scaladsl.One2OneBidiFlow$OutputTruncationException$ to be thrown, but java.lang.IllegalArgumentException was thrown. (HighLevelOutgoingConnectionSpec.scala:70)
[info]   org.scalatest.exceptions.TestFailedException:
[info]   at org.scalatest.Assertions$class.newAssertionFailedException(Assertions.scala:496)
[info]   at org.scalatest.Assertions$.newAssertionFailedException(Assertions.scala:1538)
[info]   at org.scalatest.Assertions$.checkExpectedException(Assertions.scala:1572)
[info]   at org.scalatest.words.ResultOfBeWordForAType.thrownBy(ResultOfBeWordForAType.scala:39)
[info]   at akka.http.impl.engine.client.HighLevelOutgoingConnectionSpec$$anonfun$1$$anonfun$apply$mcV$sp$3.apply$mcV$sp(HighLevelOutgoingConnectionSpec.scala:83)
[info]   at akka.http.impl.engine.client.HighLevelOutgoingConnectionSpec$$anonfun$1$$anonfun$apply$mcV$sp$3.apply(HighLevelOutgoingConnectionSpec.scala:70)
[info]   at akka.http.impl.engine.client.HighLevelOutgoingConnectionSpec$$anonfun$1$$anonfun$apply$mcV$sp$3.apply(HighLevelOutgoingConnectionSpec.scala:70)
[info]   at org.scalatest.Transformer$$anonfun$apply$1.apply$mcV$sp(Transformer.scala:22)
[info]   at org.scalatest.OutcomeOf$class.outcomeOf(OutcomeOf.scala:85)
[info]   at org.scalatest.OutcomeOf$.outcomeOf(OutcomeOf.scala:104)
[info]   at org.scalatest.Transformer.apply(Transformer.scala:22)
[info]   at org.scalatest.Transformer.apply(Transformer.scala:20)
[info]   at org.scalatest.WordSpecLike$$anon$1.apply(WordSpecLike.scala:953)
[info]   at org.scalatest.Suite$class.withFixture(Suite.scala:1122)
[info]   at akka.stream.testkit.AkkaSpec.withFixture(AkkaSpec.scala:58)
[info]   at org.scalatest.WordSpecLike$class.invokeWithFixture$1(WordSpecLike.scala:950)
[info]   at org.scalatest.WordSpecLike$$anonfun$runTest$1.apply(WordSpecLike.scala:962)
[info]   at org.scalatest.WordSpecLike$$anonfun$runTest$1.apply(WordSpecLike.scala:962)
[info]   at org.scalatest.SuperEngine.runTestImpl(Engine.scala:306)
[info]   at org.scalatest.WordSpecLike$class.runTest(WordSpecLike.scala:962)
[info]   at akka.stream.testkit.AkkaSpec.runTest(AkkaSpec.scala:58)
[info]   at org.scalatest.WordSpecLike$$anonfun$runTests$1.apply(WordSpecLike.scala:1021)
[info]   at org.scalatest.WordSpecLike$$anonfun$runTests$1.apply(WordSpecLike.scala:1021)
[info]   at org.scalatest.SuperEngine$$anonfun$traverseSubNodes$1$1.apply(Engine.scala:413)
[info]   at org.scalatest.SuperEngine$$anonfun$traverseSubNodes$1$1.apply(Engine.scala:401)
```

--- a/akka-http-core/src/main/scala/akka/http/impl/engine/client/OutgoingConnectionBlueprint.scala
+++ b/akka-http-core/src/main/scala/akka/http/impl/engine/client/OutgoingConnectionBlueprint.scala
@@ -167,7 +167,6 @@ private[http] object OutgoingConnectionBlueprint {
       // each connection uses a single (private) response parser instance for all its responses
       // which builds a cache of all header instances seen on that connection
       val parser = rootParser.createShallowCopy()
-      var methodBypassCompleted = false
       var waitingForMethod = true
 
       setHandler(methodBypassInput, new InHandler {
@@ -179,7 +178,6 @@ private[http] object OutgoingConnectionBlueprint {
         }
         override def onUpstreamFinish(): Unit =
           if (waitingForMethod) completeStage()
-          else methodBypassCompleted = true
       })
 
       setHandler(dataInput, new InHandler {
@@ -201,17 +199,16 @@ private[http] object OutgoingConnectionBlueprint {
 
       setHandler(out, eagerTerminateOutput)
 
-      val getNextMethod = () ⇒
-        if (methodBypassCompleted) completeStage()
-        else {
-          pull(methodBypassInput)
-          waitingForMethod = true
-        }
+      val getNextMethod = () ⇒ {
+        waitingForMethod = true
+        if (isClosed(methodBypassInput)) completeStage()
+        else pull(methodBypassInput)
+      }
 
       val getNextData = () ⇒ {
         waitingForMethod = false
-        if (!isClosed(dataInput)) pull(dataInput)
-        else completeStage()
+        if (isClosed(dataInput)) completeStage()
+        else pull(dataInput)
       }
 
       @tailrec def drainParser(current: ResponseOutput, b: ListBuffer[ResponseOutput] = ListBuffer.empty): Unit = {
@@ -219,13 +216,10 @@ private[http] object OutgoingConnectionBlueprint {
           if (output.nonEmpty) emit(out, output, andThen)
           else andThen()
         current match {
-          case NeedNextRequestMethod ⇒
-            e(b.result(), getNextMethod)
-          case StreamEnd ⇒
-            e(b.result(), () ⇒ completeStage())
-          case NeedMoreData ⇒
-            e(b.result(), getNextData)
-          case x ⇒ drainParser(parser.onPull(), b += x)
+          case NeedNextRequestMethod ⇒ e(b.result(), getNextMethod)
+          case StreamEnd             ⇒ e(b.result(), () ⇒ completeStage())
+          case NeedMoreData          ⇒ e(b.result(), getNextData)
+          case x                     ⇒ drainParser(parser.onPull(), b += x)
         }
       }
 

--- a/akka-http-core/src/test/scala/akka/http/impl/engine/client/HighLevelOutgoingConnectionSpec.scala
+++ b/akka-http-core/src/test/scala/akka/http/impl/engine/client/HighLevelOutgoingConnectionSpec.scala
@@ -11,16 +11,19 @@ import akka.stream.scaladsl._
 import akka.stream.testkit.AkkaSpec
 import akka.http.scaladsl.{ Http, TestUtils }
 import akka.http.scaladsl.model._
+import akka.stream.testkit.Utils
+import org.scalatest.concurrent.ScalaFutures
 
-class HighLevelOutgoingConnectionSpec extends AkkaSpec {
+class HighLevelOutgoingConnectionSpec extends AkkaSpec with ScalaFutures {
   implicit val materializer = ActorMaterializer()
+  implicit val patience = PatienceConfig(1.second)
 
   "The connection-level client implementation" should {
 
-    "be able to handle 100 pipelined requests across one connection" in {
+    "be able to handle 100 pipelined requests across one connection" in Utils.assertAllStagesStopped {
       val (_, serverHostName, serverPort) = TestUtils.temporaryServerHostnameAndPort()
 
-      Http().bindAndHandleSync(r ⇒ HttpResponse(entity = r.uri.toString.reverse.takeWhile(Character.isDigit).reverse),
+      val binding = Http().bindAndHandleSync(r ⇒ HttpResponse(entity = r.uri.toString.reverse.takeWhile(Character.isDigit).reverse),
         serverHostName, serverPort)
 
       val N = 100
@@ -32,13 +35,14 @@ class HighLevelOutgoingConnectionSpec extends AkkaSpec {
         .map { r ⇒ val s = r.data.utf8String; log.debug(s); s.toInt }
         .runFold(0)(_ + _)
 
-      Await.result(result, 10.seconds) shouldEqual N * (N + 1) / 2
+      result.futureValue(PatienceConfig(10.seconds)) shouldEqual N * (N + 1) / 2
+      binding.futureValue.unbind()
     }
 
-    "be able to handle 100 pipelined requests across 4 connections (client-flow is reusable)" in {
+    "be able to handle 100 pipelined requests across 4 connections (client-flow is reusable)" in Utils.assertAllStagesStopped {
       val (_, serverHostName, serverPort) = TestUtils.temporaryServerHostnameAndPort()
 
-      Http().bindAndHandleSync(r ⇒ HttpResponse(entity = r.uri.toString.reverse.takeWhile(Character.isDigit).reverse),
+      val binding = Http().bindAndHandleSync(r ⇒ HttpResponse(entity = r.uri.toString.reverse.takeWhile(Character.isDigit).reverse),
         serverHostName, serverPort)
 
       val connFlow = Http().outgoingConnection(serverHostName, serverPort)
@@ -64,12 +68,14 @@ class HighLevelOutgoingConnectionSpec extends AkkaSpec {
         .map { r ⇒ val s = r.data.utf8String; log.debug(s); s.toInt }
         .runFold(0)(_ + _)
 
-      Await.result(result, 10.seconds) shouldEqual C * N * (N + 1) / 2
+      result.futureValue(PatienceConfig(10.seconds)) shouldEqual C * N * (N + 1) / 2
+      binding.futureValue.unbind()
     }
 
-    "catch response stream truncation" in {
+    "catch response stream truncation" in Utils.assertAllStagesStopped {
       val (_, serverHostName, serverPort) = TestUtils.temporaryServerHostnameAndPort()
-      Http().bindAndHandleSync({
+
+      val binding = Http().bindAndHandleSync({
         case HttpRequest(_, Uri.Path("/b"), _, _, _) ⇒ HttpResponse(headers = List(headers.Connection("close")))
         case _                                       ⇒ HttpResponse()
       }, serverHostName, serverPort)
@@ -81,6 +87,7 @@ class HighLevelOutgoingConnectionSpec extends AkkaSpec {
         .runWith(Sink.head)
 
       a[One2OneBidiFlow.OutputTruncationException.type] should be thrownBy Await.result(x, 1.second)
+      binding.futureValue.unbind()
     }
   }
 }

--- a/akka-http-core/src/test/scala/akka/http/impl/engine/ws/WebsocketIntegrationSpec.scala
+++ b/akka-http-core/src/test/scala/akka/http/impl/engine/ws/WebsocketIntegrationSpec.scala
@@ -32,16 +32,18 @@ class WebsocketIntegrationSpec extends AkkaSpec("akka.stream.materializer.debug.
   "A Websocket server" must {
 
     "echo 100 elements and then shut down without error" in Utils.assertAllStagesStopped {
+
       val bindingFuture = Http().bindAndHandleSync({
         case HttpRequest(_, _, headers, _, _) ⇒
           val upgrade = headers.collectFirst { case u: UpgradeToWebsocket ⇒ u }.get
           upgrade.handleMessages(Flow.apply, None)
-      }, interface = "localhost", port = 8080)
+      }, interface = "localhost", port = 0)
       val binding = Await.result(bindingFuture, 3.seconds)
+      val myPort = binding.localAddress.getPort
 
       val N = 100
       val (response, count) = Http().singleWebsocketRequest(
-        WebsocketRequest("ws://127.0.0.1:8080"),
+        WebsocketRequest("ws://127.0.0.1:" + myPort),
         Flow.fromSinkAndSourceMat(
           Sink.fold(0)((n, _: Message) ⇒ n + 1),
           Source.repeat(TextMessage("hello")).take(N))(Keep.left))
@@ -73,18 +75,19 @@ class WebsocketIntegrationSpec extends AkkaSpec("akka.stream.materializer.debug.
         case HttpRequest(_, _, headers, _, _) ⇒
           val upgrade = headers.collectFirst { case u: UpgradeToWebsocket ⇒ u }.get
           upgrade.handleMessages(handler, None)
-      }, interface = "localhost", port = 8080)
+      }, interface = "localhost", port = 0)
       val binding = Await.result(bindingFuture, 3.seconds)
+      val myPort = binding.localAddress.getPort
 
       @volatile var messages = 0
       val (breaker, completion) =
         Source.maybe
           .viaMat {
-            Http().websocketClientLayer(WebsocketRequest("ws://localhost:8080"))
+            Http().websocketClientLayer(WebsocketRequest("ws://localhost:" + myPort))
               .atop(SslTlsPlacebo.forScala)
               // the resource leak of #19398 existed only for severed websocket connections
               .atopMat(GraphStages.bidiBreaker[ByteString, ByteString])(Keep.right)
-              .join(Tcp().outgoingConnection(new InetSocketAddress("localhost", 8080), halfClose = true))
+              .join(Tcp().outgoingConnection(new InetSocketAddress("localhost", myPort), halfClose = true))
           }(Keep.right)
           .toMat(Sink.foreach(_ ⇒ messages += 1))(Keep.both)
           .run()

